### ISSUE TITLE
Dealing with directories being larger than a single block

### ISSUE THREAD

**Issue by pierricgimmig** [MolecularMatters/raw_pdb#6]
Both raw_pdb and Llvm can't deal with pdbs that have more block indices than what fits in a single block. The DIA SDK seems to handle this case fine however. Although this rarely happens, we are dealing with pdb files that fail this `(directoryBlockCount > blockIndicesPerBlock)` validation check. This problem led us to move away from the Llvm implementation in favor of the DIA SDK's.

After some investigation on the problematic pdb files, which we unfortunately can not share, we found a [workaround](https://github.com/pierricgimmig/orbit/blob/feature/raw_pdb/third_party/raw_pdb/src/PDB.cpp).

We assume that block indices will spill to the next block(s). Inspecting the block values confirmed this assumption with our problematic pdbs. For example, a pdb file that has 824 blocks in the directory but can only fit 256 indices per block will use 4 consecutives blocks for storing indices. In our case, all the indices are consecutives, even crossing block boundaries, and the last block of indices is zero-filled after the last index.

```
directoryBlockCount = 824
blockIndicesPerBlock = 256
superBlock->blockSize = 1024
superBlock->freeBlockMapIndex = 2
superBlock->blockCount = 210763
superBlock->directorySize = 843148
superBlock->unknown = 0
superBlock->directoryIndicesBlockIndex = 210754
superBlock->directorySize / superBlock->blockSize = 823
num_indices_blocks = 4
```

The workaround currently consists of making sure that blocks are contiguous in memory (which I think is always the case thanks to the CoalescedMSFStream(?)). This way, indexing from the address of the first index block will safely overflow to the subsequent index block(s). We could add some additional validation, like making sure the indices are in a valid range, and that the last block is indeed zero-filled after the last index. In any case, the fix seems to be working for us, but I'm curious to get your thoughts on the approach. 

Do you have pdb files that trigger the check? I'll keep searching for such pdb files which I could share.

Thanks!

**Comment by MolecularMatters** [comment]
Thanks for bringing this to my attention!
I haven't seen any PDBs like this yet, it would be appreciated if you had a PDB to share.

In this case, I think the block indices **have** to be contiguous in memory:
The directory tells us which streams are made up of which blocks, which serves as an indirection and allows PDBs to store their data block-interleaved. In this case, we are reading the block directory, so there is no other indirection available that could tell us where blocks go if they were to be found at another offset in the file.

I'll fix this as soon as I get a PDB file to verify.

**Comment by pierricgimmig** [comment]
Ok, thanks for the input! So I scanned over 9K pdb files, and it seems I only have 2 that exhibit the problem, and I can't share them. I'll keep searching. At least this is documented and might help anyone running into the same problem.

**Comment by MolecularMatters** [comment]
Thanks to @rovarma who helped me find PDBs that exhibit this behaviour.
This issue has been fixed in 83516ec2d1bbedce14b0a5d5eb1f3dd1db9a3c36.

**Comment by JustasMasiulis** [comment]
> Thanks to @rovarma who helped me find PDBs that exhibit this behaviour. This issue has been fixed in [83516ec](https://github.com/MolecularMatters/raw_pdb/commit/83516ec2d1bbedce14b0a5d5eb1f3dd1db9a3c36).

IMO that's hardly a fix. If you don't want to implement the case where these blocks are non-contiguous at least do verification of that.

**Comment by MolecularMatters** [comment]
> IMO that's hardly a fix. If you don't want to implement the case where these blocks are non-contiguous at least do verification of that.

That case **is** implemented, because non-contiguous blocks are all over the place in PDB files.
In this particular case, it is being taken care of by the ContiguousMSFStream in the RawFile, see https://github.com/MolecularMatters/raw_pdb/blob/main/src/PDB_RawFile.cpp#L44.

The bug that needed fixing was the wrong assumption that the indices of all directory blocks will fit into one single block, which is not true for some PDBs.

**Comment by JustasMasiulis** [comment]
> > IMO that's hardly a fix. If you don't want to implement the case where these blocks are non-contiguous at least do verification of that.
> 
> That case **is** implemented, because non-contiguous blocks are all over the place in PDB files. In this particular case, it is being taken care of by the CoalescedMSFStream in the RawFile, see https://github.com/MolecularMatters/raw_pdb/blob/main/src/PDB_RawFile.cpp#L44.
> 
> The bug that needed fixing was the wrong assumption that the indices of all directory blocks will fit into one single block, which is not true for some PDBs.

I'm talking specifically about referenced commit and the block map. Unless I'm missing something it seems to me like you are assuming that the stream directory block map is now a contiguous array of indices that can span across multiple blocks.

```cpp
uint32_t directoryIndicesBlockIndex; // index of the first block that contains an array of indices of directory blocks, where the array itself can span several blocks
```

However to me it would seem that this definition is not correct and should be as so:

```cpp
uint32_t directoryIndicesBlockIndices[]; // indices of the blocks that contain array of indices of directory blocks, ...
```

**Comment by MolecularMatters** [comment]
Ah, I see what you mean.
I am under the assumption that the MSF SuperBlock is a fixed-size thing, so the block index to the first block that stores the directory indices really is just a single uint32_t index, and not an array of uint32_t indices.

LLVM agrees with this definition: https://llvm.org/docs/PDB/MsfFile.html#msf-superblock
More specifically, it states _"The number of ulittle32_t’s in this array is given by ceil(NumDirectoryBytes / BlockSize)."_ which is also what the current code assumes.

Of course, the LLVM format docs could be wrong, and I'd be happy to implement this in case the SuperBlock contains an array of indices.

**Comment by MolecularMatters** [comment]
I checked the PDB files I have to verify this, and they indeed store an array of indices in the SuperBlock, rather than a single index.
Thank you @JustasMasiulis for pointing that out, I will fix this accordingly.

**Comment by JustasMasiulis** [comment]
> LLVM agrees with this definition: https://llvm.org/docs/PDB/MsfFile.html#msf-superblock More specifically, it states _"The number of ulittle32_t’s in this array is given by ceil(NumDirectoryBytes / BlockSize)."_ which is also what the current code assumes.

LLVM has a lot of assumptions that may not hold up for PDBs not generated by LLVM itself.

I'd suggest either reading microsoft's horrendous code base and doing some reversing of DIA or using https://github.com/willglynn/pdb like a middle point of "good enough", as while it's probably the most accurate open source implementation it still falls short on what the actual implementation is supposed to be like (IE the address translation).

**Comment by MolecularMatters** [comment]
Fixed in 66dccb0ba06952691c1fc0b1635f87f109099ee9.

**Comment by pierricgimmig** [comment]
Very nice! Many thanks, @MolecularMatters !
